airflow/dags/unzip_and_validate_gtfs_schedule/unzip_gtfs_schedule.py

```diff
@@ -0,0 +1,205 @@
+
+    @property
+    def dt(self) -> pendulum.Date:
+        return self.ts.date()
+
+
+class GTFSScheduleFeedExtractUnzipOutcome(ProcessingOutcome):
+    zipfile_extract_path: str
+    zipfile_extract_md5hash: Optional[str]
+    zipfile_files: Optional[List[str]]
+    zipfile_dirs: Optional[List[str]]
+    extracted_files: Optional[List[str]]
+
+
+class ScheduleUnzipResult(PartitionedGCSArtifact):
+    bucket: ClassVar[str] = SCHEDULE_UNZIPPED_BUCKET
+    table: ClassVar[str] = "unzipping_results"
+    partition_names: ClassVar[List[str]] = ["dt"]
+    dt: pendulum.Date
+    outcomes: List[GTFSScheduleFeedExtractUnzipOutcome]
+
+    @property
+    def successes(self) -> List[GTFSScheduleFeedExtractUnzipOutcome]:
+        return [outcome for outcome in self.outcomes if outcome.success]
+
+    @property
+    def failures(self) -> List[GTFSScheduleFeedExtractUnzipOutcome]:
+        return [outcome for outcome in self.outcomes if not outcome.success]
+
+    def save(self, fs):
+        self.save_content(
+            fs=fs,
+            content="\n".join(o.json() for o in self.outcomes).encode(),
+            exclude={"outcomes"},
+        )
+
+
+def summarize_zip_contents(zip: zipfile.ZipFile, at: str = ""):
+    files = []
+    directories = []
+    for entry in zip.namelist():
+        if zipfile.Path(zip, at=entry).is_file():
+            files.append(entry)
+        if zipfile.Path(zip, at=entry).is_dir():
+            directories.append(entry)
+    logging.info(f"Found files: {files} and directories: {directories}")
+    return files, directories
+
+
+def process_feed_files(
+    fs, extract: GTFSScheduleFeedExtract, zip: zipfile.ZipFile, directories: List[str]
+):
+    zipfile_files = []
+    # more than one directory --> can't parse
+    if len(directories) > 1:
+        raise ValueError("Unparseable zip: Multiple directories found")
+    # directory + any other artifact at top level --> can't parse
+    if directories and len([item for item in zipfile.Path(zip, at="").iterdir()]) > 1:
+        raise ValueError("Unparseable zip: Directory and other artifacts at root level")
+    # if the only artifact at the root level of the zipfile is a directory
+    # we can treat the contents of that directory as the feed
+    if [item.is_dir() for item in zipfile.Path(zip, at="").iterdir()] == [True]:
+        feed_directory = directories[0]
+        logging.info("Top-level directory found")
+    else:
+        feed_directory = ""
+
+    for file in zipfile.Path(zip, at=feed_directory).iterdir():
+        if feed_directory:
+            file_name = os.path.join(feed_directory, file.name)
+        else:
+            file_name = file.name
+        with zip.open(file_name) as f:
+            file_content = f.read()
+        file_extract = GTFSScheduleFeedFile(
+            ts=extract.ts,
+            base64_url=extract.base64_url,
+            zipfile_path=extract.path,
+            original_filename=file.name,
```

Comment:
should this be `file_name`?

Reply:
Ope, I missed this when we were talking earlier. I guess it could go either way & depends on what we think we care about with `original_filename` -- is it just about tracking potential character substitutions to make the file GCS-safe? Or is it about tracking information about the file within the zipfile? 

The way it is now, if a file was in a directory, that directory will show up (`original_filename=dir/file.txt`). So that would be more the latter.

If we only care about slugifying for GCS, then it can be `file_name`.

Reply:
I think `original_filename` should have the directory structure maybe? So it's clear that it originally came from inside a directory within the zipfile. Is that reasonable?

Reply:
yes sorry I had confused myself. this is now doing the right thing